File: services/ai/utilities/tree_of_thoughts.ts
import { complexity_score } from '../pipeline';
import { validateAgainstSchema, SchemaType } from '../schemas/validator';

export interface ToTConfig {
  enabled: boolean;
  breadth: number; // number of branches per level
  depth: number; // maximum depth
  minComplexity: number;
  requiredPersona: string;
}

export interface ThoughtNode {
  id: string;
  content: any;
  parentId?: string;
  depth: number;
  score: number;
  valid: boolean;
  children: ThoughtNode[];
}

// Mock LLM call for thoughts
async function mockGenerateThought(prompt: string, context: string[]): Promise<any> {
  // Placeholder - would call LLM with thought generation prompt
  console.log(`[ToT] Generating thought with context: ${context.length} items`);
  return { thought: `Thought based on ${context.join(', ')}` };
}

export class TreeOfThoughts {
  private config: ToTConfig;

  constructor(config: Partial<ToTConfig> = {}) {
    this.config = {
      enabled: config.enabled ?? true,
      breadth: config.breadth ?? 3,
      depth: config.depth ?? 2,
      minComplexity: config.minComplexity ?? 0.8,
      requiredPersona: config.requiredPersona ?? 'analyst',
    };
  }

  async shouldUseToT(
    input: { userMessage: string; context?: any },
    persona: string,
    schema: SchemaType
  ): Promise<boolean> {
    if (!this.config.enabled) return false;
    if (persona !== this.config.requiredPersona) return false;
    
    const complexity = complexity_score(input);
    return complexity >= this.config.minComplexity;
  }

  async exploreThoughts(
    initialPrompt: string,
    input: { userMessage: string; context?: any },
    persona: string,
    schema: SchemaType
  ): Promise<ThoughtNode | null> {
    const shouldUse = await this.shouldUseToT(input, persona, schema);
    if (!shouldUse) return null;

    console.log(`[ToT] Starting exploration with breadth=${this.config.breadth}, depth=${this.config.depth}`);

    const root: ThoughtNode = {
      id: 'root',
      content: { initial: true },
      depth: 0,
      score: 0,
      valid: true,
      children: [],
    };

    await this.exploreLevel(root, initialPrompt, schema);

    // Find best leaf node
    const bestLeaf = this.findBestLeaf(root);
    
    console.log(`[ToT] Exploration complete. Best leaf score: ${bestLeaf?.score}`);
    
    return bestLeaf || null;
  }

  private async exploreLevel(node: ThoughtNode, prompt: string, schema: SchemaType): Promise<void> {
    if (node.depth >= this.config.depth) return;

    // Generate breadth number of thoughts
    for (let i = 0; i < this.config.breadth; i++) {
      try {
        const context = this.buildContextFromPath(node);
        const thought = await mockGenerateThought(prompt, context);
        
        // Validate thought
        const validation = validateAgainstSchema(schema, thought);
        const score = this.scoreThought(thought, validation.valid, prompt);
        
        const childNode: ThoughtNode = {
          id: `${node.id}-${i}`,
          content: thought,
          parentId: node.id,
          depth: node.depth + 1,
          score,
          valid: validation.valid,
          children: [],
        };
        
        node.children.push(childNode);
        
        // Recursively explore if valid and high-scoring
        if (validation.valid && score > 0.7) {
          await this.exploreLevel(childNode, prompt, schema);
        }
      } catch (error) {
        console.warn(`[ToT] Failed to generate thought ${i} at depth ${node.depth}:`, error);
      }
    }
  }

  private buildContextFromPath(node: ThoughtNode): string[] {
    const context: string[] = [];
    let current: ThoughtNode | undefined = node;
    
    while (current) {
      if (current.content && typeof current.content === 'object') {
        context.unshift(JSON.stringify(current.content));
      }
      current = current.parentId ? this.findNodeById(current.parentId, node) : undefined;
    }
    
    return context;
  }

  private findNodeById(id: string, root: ThoughtNode): ThoughtNode | undefined {
    if (root.id === id) return root;
    for (const child of root.children) {
      const found = this.findNodeById(id, child);
      if (found) return found;
    }
    return undefined;
  }

  private scoreThought(thought: any, isValid: boolean, objective: string): number {
    let score = 0;
    
    // Schema validity (50% weight)
    score += isValid ? 0.5 : 0;
    
    // Objective coverage (50% weight) - placeholder logic
    const thoughtStr = JSON.stringify(thought).toLowerCase();
    const objectiveWords = objective.toLowerCase().split(/\s+/);
    const coverage = objectiveWords.filter(word => thoughtStr.includes(word)).length / objectiveWords.length;
    score += coverage * 0.5;
    
    return Math.min(score, 1);
  }

  private findBestLeaf(root: ThoughtNode): ThoughtNode | null {
    const leaves: ThoughtNode[] = [];
    
    function collectLeaves(node: ThoughtNode) {
      if (node.children.length === 0) {
        leaves.push(node);
      } else {
        node.children.forEach(collectLeaves);
      }
    }
    
    collectLeaves(root);
    
    // Return highest scoring leaf
    return leaves.reduce((best, current) => 
      current.score > best.score ? current : best
    );
  }
}

// Singleton instance
export const treeOfThoughts = new TreeOfThoughts();

File: services/python-ai/app/ai_legacy/utilities/tree_of_thoughts.py
from typing import Any, Dict, List, Optional

from ..pipeline import complexity_score
from ..schemas.validator import SchemaType, validate_against_schema


class TreeOfThoughts:
    def __init__(self, enabled: bool = True, breadth: int = 3, depth: int = 2, min_complexity: float = 0.8, required_persona: str = "analyst") -> None:
        self.config = {
            "enabled": enabled,
            "breadth": breadth,
            "depth": depth,
            "minComplexity": min_complexity,
            "requiredPersona": required_persona,
        }

    async def should_use_tot(self, input_payload: Dict[str, Any], persona: str, schema: SchemaType) -> bool:
        if not self.config["enabled"]:
            return False
        if persona != self.config["requiredPersona"]:
            return False
        complexity = complexity_score(input_payload)
        return complexity >= self.config["minComplexity"]

    async def explore_thoughts(self, initial_prompt: str, input_payload: Dict[str, Any], persona: str, schema: SchemaType) -> Optional[Dict[str, Any]]:
        if not await self.should_use_tot(input_payload, persona, schema):
            return None
        root = {"id": "root", "content": {"initial": True}, "depth": 0, "score": 0, "valid": True, "children": []}
        await self._explore_level(root, initial_prompt, schema)
        return self._find_best_leaf(root)

    async def _explore_level(self, node: Dict[str, Any], prompt: str, schema: SchemaType) -> None:
        if node["depth"] >= self.config["depth"]:
            return
        for i in range(self.config["breadth"]):
            thought = await self._mock_generate_thought(prompt, self._build_context_from_path(node))
            validation = validate_against_schema(schema, thought)
            score = self._score_thought(thought, validation["valid"], prompt)
            child = {
                "id": f"{node['id']}-{i}",
                "content": thought,
                "parentId": node["id"],
                "depth": node["depth"] + 1,
                "score": score,
                "valid": validation["valid"],
                "children": [],
            }
            node["children"].append(child)
            if validation["valid"] and score > 0.7:
                await self._explore_level(child, prompt, schema)

    async def _mock_generate_thought(self, prompt: str, context: List[str]) -> Any:
        return {"thought": f"Thought based on {', '.join(context)}"}

    def _build_context_from_path(self, node: Dict[str, Any]) -> List[str]:
        context: List[str] = []
        current: Optional[Dict[str, Any]] = node
        while current:
            if isinstance(current.get("content"), dict):
                context.insert(0, str(current["content"]))
            parent_id = current.get("parentId")
            current = self._find_node_by_id(parent_id, node) if parent_id else None
        return context

    def _find_node_by_id(self, node_id: Optional[str], root: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        if not node_id:
            return None
        if root["id"] == node_id:
            return root
        for child in root.get("children", []):
            found = self._find_node_by_id(node_id, child)
            if found:
                return found
        return None

    def _score_thought(self, thought: Any, is_valid: bool, objective: str) -> float:
        score = 0.5 if is_valid else 0.0
        thought_str = str(thought).lower()
        objective_words = objective.lower().split()
        coverage = len([word for word in objective_words if word in thought_str]) / max(len(objective_words), 1)
        score += coverage * 0.5
        return min(score, 1.0)

    def _find_best_leaf(self, root: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        leaves: List[Dict[str, Any]] = []

        def collect(node: Dict[str, Any]) -> None:
            if not node.get("children"):
                leaves.append(node)
                return
            for child in node.get("children", []):
                collect(child)

        collect(root)
        if not leaves:
            return None
        return max(leaves, key=lambda leaf: leaf.get("score", 0))


tree_of_thoughts = TreeOfThoughts()
